lib/mongo/socket.rb

```diff
@@ -123,6 +123,7 @@ def initialize(family)
     def read(length)
       handle_errors do
         data = read_from_socket(length)
+        raise IOError unless data.length == length
```

Comment:
The reason the auth tests fail is because some authentication protocols receive binary data of length 0 at the end of the conversation. This is noted in the auth specs:

```
S: {conversationId: 1, payload: BinData(0,""), done: true, ok: 1}
Note
There is an extra round trip due to an implementation decision on the server. This is accomplished by sending no bytes back to the server for what is effectively a no-op.
```

With this PR, an error would be raised when this empty binary data is attempted to be read off the socket.
I'll add a check if (data.length == 0 && length != 0)
In other words, I'll check if you read data of length 0 and explicitly asked to read data of length 0.